core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala

```diff
@@ -0,0 +1,156 @@
+    Utils.executeCommand(
+      Seq(s"$sparkHome/bin/spark-class",
+        exceptionThrower.getClass.getCanonicalName.dropRight(1), // drops the "$" at the end
+        exitOnUncaughtException.toString),
+      new File(sparkHome),
+      Map("SPARK_TESTING" -> "1", "SPARK_HOME" -> sparkHome))
+  }
+
+  test("SPARK-30310: Test uncaught RuntimeException, exitOnUncaughtException = true") {
+    val process = getThrowerProcess(RuntimeExceptionThrower, exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.UNCAUGHT_EXCEPTION)
+  }
+
+  test("SPARK-30310: Test uncaught RuntimeException, exitOnUncaughtException = false") {
+    val process = getThrowerProcess(RuntimeExceptionThrower, exitOnUncaughtException = false)
+    assert(process.waitFor == 0)
+  }
+
+  test("SPARK-30310: Test uncaught OutOfMemoryError, exitOnUncaughtException = true") {
+    val process = getThrowerProcess(OutOfMemoryErrorThrower, exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+  test("SPARK-30310: Test uncaught OutOfMemoryError, exitOnUncaughtException = false") {
+    val process = getThrowerProcess(OutOfMemoryErrorThrower, exitOnUncaughtException = false)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException, exitOnUncaughtException = true") {
+    val process = getThrowerProcess(SparkFatalExceptionThrower, exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.UNCAUGHT_EXCEPTION)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException, exitOnUncaughtException = false") {
+    val process = getThrowerProcess(SparkFatalExceptionThrower, exitOnUncaughtException = false)
+    assert(process.waitFor == 0)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException (OOM), exitOnUncaughtException = true") {
+    val process = getThrowerProcess(SparkFatalExceptionWithOOMThrower,
+      exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException (OOM), exitOnUncaughtException = false") {
+    val process = getThrowerProcess(SparkFatalExceptionWithOOMThrower,
+      exitOnUncaughtException = false)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+}
+
+// a thread that uses SparkUncaughtExceptionHandler, then throws the throwable
+class ThrowableThrowerThread(t: Throwable,
+    exitOnUncaughtException: Boolean) extends Thread {
+  override def run() {
+    Thread.setDefaultUncaughtExceptionHandler(
+      new SparkUncaughtExceptionHandler(exitOnUncaughtException))
+    throw t
+  }
+}
+
+// Objects to be invoked by spark-class for different Throwable types
+// that SparkUncaughtExceptionHandler handles.  spark-class will exit with
+// exit code dictated by either:
+// - SparkUncaughtExceptionHandler (SparkExitCode)
+// - main() (0, or -1 when args is empty)
+
+object RuntimeExceptionThrower {
```

Comment:
All objects have same implementation except which exception to create and pass. Why not consolidating into one object, and additional argument to differentiate the exception? Define a new Enumeration for this and pass it as String for additional argument.

Reply:
Thanks for the suggestion!   Modified as such.